docs/docs/policy-language.md

```diff
@@ -200,6 +200,66 @@ escape special characters.
 
 A simple example is a regex to match a valid Rego variable. With a regular string, the regex is `"[a-zA-Z_]\\w*"`, but with raw strings, it becomes `` `[a-zA-Z_]\w*` ``.
 
+### String Interpolation
+
+Runtime data can be incorporated into a string through string interpolation. An interpolated string is composed of a template-string containing zero or more template-expressions.
+The `$` character identifies a template-string, and can be used with regular double-quoted strings (`$"hello"`), and backtick-quoted raw strings (`` $`hello` ``).
+
+A template-expression is enclosed in curly-braces (`{`,`}`), and must contain a single expression that evaluate to a value, e.g.:
+
+- Primitive values: `$"{1} {2.3} {"foo"} {false} {null}"`
+- Composite values: `$"{[true, false]} {{1, 2}} {{"a": "b"}}"`
+- Variables: `x = "foo"; a = "$"{x}"`
+- References: `"$"{input.x} {data.y}"`
+- Comprehensions: `$"{[x | ...]} {{x | ...}} {{x: y | ...}}"`
+
+```rego
+package interpolation
+
+username := "Alice"
+
+a := $"Hello {username}!"
+```
+
+<RunSnippet command="data.interpolation.a"/>
+
+#### Undefined values
+
+`Undefined` values don't halt evaluation of template-strings. If a template-expression evaluates to an `undefined` value,
+the string `"<undefined>"` will be emitted instead. This means string interpolation is safe to use in cases where a string result is
+always expected, but not all expression values are guaranteed at evaluation time.
+
+```rego
+package interpolation
+
+allowed_roles := ["admin", "employee"]
+
+default role := "guest"
+role := input.role
+
+deny contains $"User {input.username}'s role was '{role}', but must be one of {allowed_roles}" if {
+    not role in allowed_roles
+}
+```
+
+<RunSnippet command="data.interpolation.deny"/>
+
+Compared to the `sprintf` [built-in function](#built-in-functions), not halting evaluation on `undefined` values make interpolated strings less error-prone, and is therefore the recommended alternative.
```

Comment:
Can we call out that it's about the `input.username` specifically? The other values can't be undefined, but let's just be super clear. I think it's a crucial point.

Reply:
> had we instead used `sprintf` in the rule head, the rule would have failed to evaluate even though `input.username` is inconsequential to the logic in the rule's body.

Can we do a side-by-side here? Something like
```rego
sprint("User %s's role was '%s', but must be one of %v", [input.username, role, allowed_roles]) => undefined

$"User {input.username}'s role was '{role}', but must be one of {allowed_roles|}" => "User <undefined>'s role was 'viewer', but must be one of [\"admin\", \"guest\"]"
```

I think it's important, and an example is easier to grok than a paragraph of text.

Reply:
What's in the changelog below looks good, too.